src/libproc_macro/lib.rs

```diff
@@ -212,12 +214,106 @@ impl Span {
         ::__internal::with_sess(|(_, mark)| Span(mark.expn_info().unwrap().call_site))
     }
 
+    /// The original source file into which this span points.
+    #[unstable(feature = "proc_macro", issue = "38356")]
+    pub fn source_file(&self) -> Option<SourceFile> {
+        // If this span's token was produced by a macro
+        if self.0.ctxt() != SyntaxContext::empty() {
+            return None
```

Comment:
I'm not sure we need to return `None` here -- if the span came from a macro definition then it would just return the macro definition's source file. We could provide a method `span.is_expanded()` (and ideally more later) to expose the span's expansion history explicitly.